### PR TITLE
Mileston-8/Blog app - add devise

### PR DESCRIPTION
**Pull request by edwardoluyara**
Here in this milestone I have was able to implement the following:
- Remove `current_user` method in `ApplicationController`, devise will provide us with one.
- [Install and setup devise](https://github.com/heartcombo/devise#getting-started).
  - Can register a new user.
  - User logs in with a combination of email and password.
  - Hashed password should be stored in the database.
  - Ask for confirmation of email.
  - Can reset password.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class User < ApplicationRecord
+  devise :database_authenticatable, :registeratable, :recoverable, :rememberable, :validatable, :confirmable
```
Comment (ShoiraTa) [inline app/models/user.rb]: - [ ] You have a type here should be `registeratable` => `registerable` kindly fix it

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,5 +1,19 @@
 class ApplicationController < ActionController::Base
-  def current_user
-    User.first
+  helper_method :current_user
+
+  protect_from_forgery with: :exception
+
+  befor_action :update_params_access, if: :devise_controller?
```
Comment (ShoiraTa) [inline app/controllers/application_controller.rb]: - [ ] You have a typo here `befor_action` => `before_action` kindly fix it

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,5 +1,19 @@
...
+  protected
+
+  def update_params_access
+    devise_param_sanitizer.permit(:sign_up) do |user|
+      user.permit(:name, :posts_counter, :surname, :email, :password, :password_confirmation)
+    end
+
+    devise_param_sanitizer.permit(:account_update) do |user|
+      user.permit(:name, :surname, :email, :password, :current_password)
+    end
```
Comment (ShoiraTa) [inline app/controllers/application_controller.rb]: - [ ] I cannot create a user, getting an error, `method_missing': undefined method email`. I checked your schema, there is no email field for users table, but users email it is being called and used thoughout the application. Kindly update user table add email and password fields and make sure that you can create a new user.

File: db/migrate/20220512033818_add_role_to_user.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUser < ActiveRecord::Migration[7.0]
+  def change
+    add_column :users, :role, :string, default: 'user'
+  end
+end
```
Comment (ShoiraTa) [inline db/migrate/20220512033818_add_role_to_user.rb]: - [ ] Within this milestone you should add divice for users table. So far you have succesfully installed divice but did not generate migration file for users. 
- [ ] to do that kindly run , `rails g devise user `

then open migration file and uncomment 
necessary fields like confirmable etc. 

then run `db:migrate`

After, check your users table it should get new password, email and other new fields 

- [ ] Finally, please make sure that you can create a new user using `rails c`.
- [ ] You can check divice documentation and watch videos on Youtube if you get stuck or keep getting errors.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,5 +1,19 @@
 class ApplicationController < ActionController::Base
-  def current_user
-    User.first
+  helper_method :current_user
```
Comment (ShoiraTa) [inline app/controllers/application_controller.rb]: - [ ] The `current_user` method in `ApplicationController` should be  removed. Devise provides a method to do this.

**Review by ShoiraTa — CHANGES_REQUESTED** [review]
Hi @team,

:thumbsup: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.

**First let me to point out:**

- Great commit history! ✔️ 
- You passed all the Linters! ✔️ 
- Professional pull request! ✔️ 

Keep on coding, good job so far! :clap:

There are some issues that you still need to work on to go to the next project but you are almost there!



 ## Required Changes  :recycle: 

_Check the comments under the review._ 

***
Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear by mentioning my Github username.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

* * *
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb [hunk 6]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (SajjadAhmad14) [inline app/models/user.rb:4]: The user is not able to sign up using email and password as there is no user interface for that. Kindly provide a link to sign up and login form on the home page which should follow this [wireframe](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md).

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Edward,
### Great work on making the changes requested by a previous reviewer 👏 
### You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.
## To Improve 🙏

- [ ] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by edwardoluyara** [comment]
Hello @SajjadAhmad14, I have the login page as shown:
CaptureI

**Review by AshakaE — APPROVED** [review]
Hi @oluyaratosin123 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**